crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4640,10 +4650,413 @@ mod tests {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            e = 1
+            f = 2
+
+            try:
+                x
+            except NameError:
+                e = 42
+            except TypeError:
+                f = 43
+            except RuntimeError:
+                f = 44
+            except OSError:
+                g = 45
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(&db, "src/a.py", "e", "Literal[1, 42]");
+        assert_public_ty(&db, "src/a.py", "f", "Literal[2, 43, 44]");
+        assert_public_ty(&db, "src/a.py", "g", "Unbound | Literal[45]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_definitions_with_finally_branch() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x
+            except NameError:
+                e = 42
+            finally:
+                e = 44
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // The `finally` block is always executed:
+        assert_public_ty(&db, "src/a.py", "e", "Literal[44]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn exhaustive_except_handler_definitions_with_else_branch() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 42
+            except NameError:
+                x = 43
+            except:
+                x = 44
+            else:
+                x = 45
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // Either one of the `except` branches or the `else` branch must have been taken;
+        // the `x` definition in the `try` block must have been overridden after completion
+        // of the entire `try`/`except` block:
+        assert_public_ty(&db, "src/a.py", "x", "Literal[43, 44, 45]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_multiple_definitions_in_try_block_with_exhaustive_excepts(
+    ) -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 1
+                y = 2
+                z = 3
+            except NameError:
+                x = 11
+            except:
+                x = 12
+                y = 22
+            else:
+                x = 13
+                y = 23
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // TODO: What would be great in these assertions would be
+        // if we could assert the types of values *within* the `else` block.
+        // (Not possible with our current testing infrastructure, sadly.) --Alex
```

Comment:
It's possible, just a bit of a pain. But on second thought, it's less of a pain than I'd originally thought. If you have a separate variable (let's say `reveal_x`) just for the purpose of inspecting a value in a nested location, and initialize it up top to a value nothing else is initialized to (`reveal_x = 0`), and then at the nested location where you want to inspect `x` you do `reveal_x = x`), then if you inspect `reveal_x` at the end you'll get a union of `Literal[0]` and the value of `x` where you wanted to inspect it. It's not quite as clear as a `reveal_type` in the new testing framework will be, but it's not bad. Up to you if you find that nicer than a TODO comment, if you're editing this PR anyway.

Reply:
We now have `reveal_type`, so this is more easily doable, even though we don't have the full new testing framework yet